[GSoC 2025 - Acceptance Tests] M2.7 (Fixes part of #22716) - Implement Acceptance tests for Curriculum Admin (CA)

jayam04 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of https://github.com/oppia/oppia/issues/22716.
2. This PR does the following:
    - Implements acceptance tests for Curriculum Admin. (CA from [CUJv3 Doc](https://docs.google.com/document/d/1D7kkFTzg3rxUe3QJ_iPlnxUzBFNElmRkmAWss00nFno/edit?tab=t.0))

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [ ] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [ ] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ ] I have written tests for my code.
- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

CI Run: https://github.com/jayam04/oppia/actions/runs/17578057448/job/49927814320?pr=77
Each test was ran 25 times in mobile and desktop.

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

jayam04 committed:
First commit

jayam04 committed:
Few changes

jayam04 committed:
Updates acceptance.json

jayam04 committed:
Merge remote-tracking branch 'upstream/develop' into m2.7

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-edit-and-delete-a-classroom.spec.ts:
```diff
@@ -0,0 +1,161 @@
+  }, 480000);
+
+  it('should be able to create a new classroom', async function () {
+    await curriculumAdmin.createNewClassroom('Math', 'math');
```
Does this use this URL to create classroom: https://www.oppiatestserver.org/classroom-admin?

jayam04 replied:
Yes. But it uses `/classroom-admin`

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-edit-and-delete-a-classroom.spec.ts:
```diff
@@ -0,0 +1,161 @@
+      'Course details',
+      'Topic list intro'
+    );
+
```
The part of `Classroom’s Name, URL Fragment, Teaser Text, Topic List Intro, Course Details, Thumbnail picture, Banner image, Publication Status, Diagnostic Test Status, Topics and their prerequisites should show and be edible` is missing.

jayam04 replied:
I added suggestion to actually edit them rather than just checking if they are editable. However, it seems someone accepted suggestion to edit them but didn't delete to old check. 

I have removed this check from the doc directly. As, we already tested they are editable by actually editing them. Let if know if this makes sense, or I should re-add this check.

jayam04 committed:
Fixes UserFactory errors

jayam04 committed:
Fix: Removes a test step which is not as expected, fix a function

jayam04 committed:
Fix: Updates wrong selector

jayam04 committed:
Fix: Fixes all issues in create, edit, and delete classroom

jayam04 committed:
Fix: Updates checks to check for element in new tab

jayam04 committed:
Update full stack tests: Run each test 25 times with no video recording

jayam04 committed:
Reverts full_stack_tests.yml

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-publish-unpublish-and-delete-topic-and-skill.spec.ts:
Add `CA. Create, publish, unpublish, and delete a topic and a skill.` to the fileoverview.

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-publish-unpublish-and-delete-topic-and-skill.spec.ts:
```diff
@@ -51,59 +52,84 @@ describe('Curriculum Admin', function () {
+    const newPage = await curriculumAdmin.clickOnElementAndGetNewPage('Save');
+    await curriculumAdmin.expectToBeInSkillEditorPage(newPage);
+  });
+
```
The part of checking "publish a skill" is missing. It should be a new it block.

jayam04 replied:
Sorry, it seems someone rejected all of suggestion of where I have deleted something. We don't need to publish a skill like topic (as given in CUJv3 Doc). The skill is already published as we create it. In fact, we don't have option to publish a skill by add write description of changes.

I am deleting the row directly, please let me know if I should undo this.

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-publish-unpublish-and-delete-topic-and-skill.spec.ts:
```diff
@@ -51,59 +52,84 @@ describe('Curriculum Admin', function () {
+    await curriculumAdmin.expectTopicNotInTopicsAndSkillDashboard(
+      'Test Topic 1'
+    );
+  });
```
There is another expectation after deleting a topic: `The respective links for Topic should return a 404 error`. If this is not correct, update the CUJ doc.

jayam04 replied:
I am assuming you are talking about last row "delete a topic". Again I deleted it in suggestion, but was rejected. I am deleting it directly.

imchristie commented on core/tests/puppeteer-acceptance-tests/specs/curriculum-admin/create-publish-unpublish-and-delete-topic-and-skill.spec.ts:
```diff
@@ -51,59 +52,84 @@ describe('Curriculum Admin', function () {
+    await curriculumAdmin.deleteSkill('Test Skill 1');
+    await curriculumAdmin.expectSkillNotInTopicsAndSkillsDashboard(
+      'Test Skill 1'
+    );
```
The part of `The respective links for skill should return a 404 error` is missing.

jayam04 replied:
@chris7716 Done. For this step, I have added steps in line 132 and 133. Let me know if that's okay.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/common/puppeteer-utils.ts:
```diff
@@ -483,6 +484,41 @@ export class BaseUser {
+    const newPagePromise: Promise<Page> = new Promise<Page>(resolve =>
+      this.browserObject.once('targetcreated', async target => {
+        const newPage = await target.page();
+        await newPage.bringToFront();
```
Is this line a must as well as the line below?

jayam04 replied:
Removed line `await page.bringToFront();`. The below line is necessary as the page returned is used in line 500.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/common/puppeteer-utils.ts:
```diff
@@ -483,6 +484,41 @@ export class BaseUser {
+  async clickOnElementAndGetNewPage(selector: string): Promise<Page> {
+    const newPagePromise: Promise<Page> = new Promise<Page>(resolve =>
+      this.browserObject.once('targetcreated', async target => {
+        const newPage = await target.page();
```
It would be better if you rename this variable so it's not the same as the one in line 501

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/common/user-factory.ts:
```diff
@@ -42,20 +42,23 @@ import {ModeratorFactory} from '../user/moderator';
 
 /**
- * Mapping of user roles to their respective function class.
+ * Mapping of user roles to their respective function class.ser
```
typo?

jayam04 replied:
Yes. Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/common/user-factory.ts:
```diff
@@ -267,7 +271,19 @@ export class UserFactory {
     activeUsers.push(user);
 
-    return await UserFactory.assignRolesToUser(user, roles, args);
+    return (await UserFactory.assignRolesToUser(
```
Why change to list out all the user roles?

jayam04 replied:
This is becuase I extended CurriculumAdmin from TopicManager instead of BaseUser.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -294,7 +296,257 @@ const mobileSaveSkillButton = '.e2e-test-mobile-save-skill-changes';
+const classroomTopicNameSelector = '.e2e-test-classroom-topic-name';
+const movableClassroomTileSelector = '.e2e-test-movable-classroom-tile';
+
+export class CurriculumAdmin extends TopicManager {
```
Why curriculum admin is under topic manger instead of base user?

jayam04 replied:
Many TopicManager functions are used in CurriculumAdmin. So, I extended CurriculumAdmin from TopicManager instead of duplicating them. Should I change to BaseUser instead?

imchristie replied:
Change back to BaseUser. It could cause confusion of extending Curriculum Admin from Topic Manager.

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -294,7 +296,257 @@ const mobileSaveSkillButton = '.e2e-test-mobile-save-skill-changes';
+      await this.page.mouse.move(targetCenter.x, targetCenter.y, {steps: 10});
+      await this.page.mouse.up();
+
+      await this.page.waitForTimeout(100);
```
Any other better way other than hardcoded the timeout?

jayam04 replied:
Removed.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -1572,6 +1822,20 @@ export class CurriculumAdmin extends BaseUser {
+   */
+  async expectUnpublishTopicButtonToBeVisible(): Promise<void> {
+    if (this.isViewportAtMobileWidth()) {
+      await this.clickOn(mobileOptionsSelector);
```
Missing pre-check

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -1942,24 +2207,37 @@ export class CurriculumAdmin extends BaseUser {
     teaserText: string,
     topicListIntro: string,
-    courseDetails: string
+    courseDetails: string,
```
Missing function description

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -1971,12 +2249,70 @@ export class CurriculumAdmin extends BaseUser {
+    topicName: string,
+    prerequisiteTopic: string | null
+  ): Promise<void> {
+    const topicPrerequisitesContainerSelector =
```
Store this variable as constant.

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -1971,12 +2249,70 @@ export class CurriculumAdmin extends BaseUser {
+    const topicBox = await this.expectClassroomToContainTopic(topicName);
+
+    const prerequisiteInputElement =
+      await topicBox.waitForSelector('mat-form-field');
```
Store this variable as constant.

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -1971,12 +2249,70 @@ export class CurriculumAdmin extends BaseUser {
   async addTopicToClassroom(
     classroomName: string,
-    topicName: string
+    topicName: string,
```
Missing function description.

jayam04 replied:
Done.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/curriculum-admin.ts:
```diff
@@ -2498,6 +2866,17 @@ export class CurriculumAdmin extends BaseUser {
+    const selector = this.isViewportAtMobileWidth()
+      ? createNewSkillMobileButton
+      : createNewSkillButtonInSkillDashboardSelector;
+    await this.clickOn(selector);
```
Missing pre-check

jayam04 replied:
Done.

imchristie reviewed (changes requested):
completed full review.

jayam04 committed:
Addresses Comments (1)

jayam04 committed:
Addressed remaining comments

imchristie reviewed (commented):
Only one unresolved comment: Move curriculum admin back to under base user instead of topic manager.

jayam04 committed:
Addresses Comments

jayam04 committed:
Updates Puppeteer not to follow new tab

imchristie reviewed (approved):
LGTM.

imchristie commented:
@Vir-8 Could you review on behalf of oppia/lace-frontend-reviewers and oppia/dev-workflow-reviewers? @U8NWXD is out on Sep 6-19

jayam04 committed:
Fix: Diff Screenshot

jayam04 committed:
Fix: expectToBeInTopicEditor

jayam04 committed:
Fix: Update screenshots

jayam04 committed:
Potential Fix: removeLessonFromPlayLater

jayam04 committed:
Potential Fix: updateProfilePicture

jayam04 committed:
Potential Fix: addBasicAlgebraQuestionToSkill

jayam04 commented:
Filed Flake: https://github.com/oppia/oppia/issues/21200#issuecomment-3291788785

Vir-8 reviewed (approved):
LGTM for codeowner files

jayam04 committed:
Renames a test suite, fixes a flake

jayam04 committed:
Merge remote-tracking branch 'upstream/develop' into m2.7

jayam04 committed:
Refactor: Moves old files back to specs/ folder

jayam04 committed:
Revert "Refactor: Moves old files back to specs/ folder"

This reverts commit eb7b1b766a7fa0b2873ee26777bddf9476bb88db.

jayam04 commented:
Filed Flake:
- https://github.com/oppia/oppia/issues/23420

jayam04 committed:
Refactored old tests back to specs folder

jayam04 committed:
Update create-subtopic-study-guide-with-multiple-sections.spec.ts

jayam04 committed:
Fix interaction tab selection in exploration editor

The commit removes redundant tab switching code since the functionality
was already handled by changeTabInInteractionSelectionModal method.

jayam04 committed:
Merge remote-tracking branch 'upstream/develop' into m2.7

jayam04 committed:
Fix: Lint errors

jayam04 committed:
Adds curriculum-admin/create-subtopic-study-guide-with-multiple-sections
to broken suites

jayam04 committed:
Update test timeout for study guide setup

jayam04 committed:
Revert "Adds curriculum-admin/create-subtopic-study-guide-with-multiple-sections"

This reverts commit c7fdc6816edb51df850fd330d1e8629fd091276f.

jayam04 committed:
Increase timeout for curriculum admin subtopic test

jayam04 commented:
Filed flake: https://github.com/oppia/oppia/issues/21200#issuecomment-3334687781
